### PR TITLE
Implement MSC3912: Relation-based redactions

### PR DESCRIPTION
**Pull request by weeman1337**
SDK support for [MSC3912](https://github.com/matrix-org/matrix-spec-proposals/pull/3912)

Should be reviewed commitwise, because it contains a heavy refactoring of the client spec.

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Implement MSC3912: Relation-based redactions ([\#2954](https://github.com/matrix-org/matrix-js-sdk/pull/2954)).

### CODE REVIEW

**Review by gsouquet — COMMENTED** [review]
Changes related to MSC3912 make sense

Glancing at the test changes, seem to make relative sense, but it would be interesting to have a better description of what the test changes are actually doing given the very large diff.

**Comment by weeman1337** [comment]
> Glancing at the test changes, seem to make relative sense, but it would be interesting to have a better description of what the test changes are actually doing given the very large diff.

The test did set up one specific client. I just moved the current client setup into a `describe` block. This way you can add more blocks with other initial configurations. Alternatively it could be a separate file for reach client setup, because the client test is already huge. Would also include some refactoring in form of extraction of some test utilities.

**Comment by weeman1337** [comment]
@gsouquet did a force push because everything changed based on the feedback or your previous review. No crazy client refactoring any more :slightly_smiling_face: . Also using `features.ts` like you suggested :+1:

**Review by gsouquet — APPROVED** [review]

File: src/@types/requests.ts [hunk 1]
```diff
@@ -47,7 +47,13 @@ export interface IJoinRoomOpts {
 
 export interface IRedactOpts {
     reason?: string;
-    // also delete related events, if the server supports it (MSC3912)
+    /**
+     * Also try to delete related events of the relation types in this array.
```
Comment (richvdh) [inline src/@types/requests.ts:51]: I think we should be more explicit about what this actually means:

```suggestion
     * Whether events related to the redacted event should be redacted.
     *
     * If specified, then any events which relate to the event being redacted with
     * any of the relationship types listed will also be redacted.
     *
```

**Review by richvdh — APPROVED** [review]

File: src/@types/requests.ts (hunk 1 above)
Comment (weeman1337) [reply src/@types/requests.ts]: Thanks for your suggestion :+1: Updated accordingly

**Pull request merged**